src/reference/mod.rs

```diff
@@ -0,0 +1,367 @@
+//! statements that construct **_graph nodes_**, link them with **_edges_**, and annotate both with
+//! **_attributes_**.
+//!
+//! [query]: https://tree-sitter.github.io/tree-sitter/using-parsers#pattern-matching-with-queries
+//! [captures]: https://tree-sitter.github.io/tree-sitter/using-parsers#capturing-nodes
+//!
+//! Comments start with a semicolon, and extend to the end of the line.
+//!
+//! To execute a graph DSL file against a concrete syntax tree, we execute each stanza in the graph
+//! DSL file _in order_.  For each stanza, we identify each place where the concrete syntax tree
+//! matches the query pattern.  For each of these places, we end up with a different set of nodes
+//! assigned to the query pattern's captures.  We execute the block of statements for each of these
+//! capture assignments, creating any nodes, edges, or attributes mentioned in the block.
+//!
+//! For instance, the following stanza would match all of the identifiers in our example syntax
+//! tree:
+//!
+//! ``` tsg
+//! (identifier) @id
+//! {
+//!   ; Some statements that will be executed for each identifier in the
+//!   ; source file.  You can use @id to refer to the matched identifier
+//!   ; node.
+//! }
+//! ```
+//!
+//! # Expressions
+//!
+//! The value of an expression in the graph DSL can be any of the following:
+//!
+//!   - a boolean
+//!   - a string
+//!   - an integer (unsigned, 32 bits)
+//!   - a reference to a syntax node
+//!   - a reference to a graph node
+//!   - an ordered list of values
+//!   - an unordered set of values
+//!
+//! The boolean literals are spelled `#true` and `#false`.
+//!
+//! String constants are enclosed in double quotes.  They can contain backslash escapes `\\`, `\0`,
+//! `\n`, `\r`, and `\t`:
+//!
+//!   - `"a string"`
+//!   - `"a string with\na newline"`
+//!   - `"a string with\\a backslash"`
+//!
+//! Integer constants are encoded in ASCII decimal:
+//!
+//!   - `0`
+//!   - `10`
+//!   - `42`
+//!
+//! # Syntax nodes
+//!
+//! Syntax nodes are identified by tree-sitter query captures (`@name`).  For instance, in our
+//! example stanza, whose query is `(identifier) @id`, `@id` would refer to the `identifier` node
+//! that the stanza matched against.
+//!
+//! # Graph nodes
+//!
+//! Each node in the output graph is identified by a syntax node and a list of **_tag names_**.
+//! Graph nodes are automatically created when they are referenced.  Graph node expressions have
+//! one of the following formats:
+//!
+//!   - `[syntax node].[tag name]`
+//!   - `[graph node].[tag name]`
+//!
+//! For instance, all of the following are valid graph node expressions in our example stanza:
+//!
+//!   - `@id.node`
+//!   - `@id.node.another`
+//!
+//! For both expressions, the graph node's syntax node would be whichever node was captured by
+//! `@id` in the stanza's query.  They would have `node` and `node.another`, respectively, as their
+//! tag names.
```

Comment:
I read this as meaning that the two `.`s in `@id.node.another` are different, the first delineating the syntax node `@id` from its tag, and the second just being a part of the tag `node.another`.

Would it make sense to perhaps exclude `.` as an allowed character in tags, in case you want to assign a different meaning to `@id.node.another` in the future? 

Also, there should probably be a specification of what kinds of characters are permitted in tag names.

Reply:
Good suggestions, thanks!

Reply:
The current implementation of the DSL is such that tags are lists of strings, not just single strings, so that if you have a tagged node `(n, ts)`, the syntax for that is `"@" ++ concat (map ("."++) ts)`. So you get `.` as the separator between each of the tag parts.

The general idea behind this is that the tags are intended to kind of.. associated nodes with other nodes in a convenient way. For instance, you might have a value associated with a node `@node.value` and that value might have a scope that it's relative to `@node.value.scope`, or so forth.

It can be removed but it would be a breaking change wrt current code.